src/platform/ESP32/BLEManagerImpl.cpp

```diff
@@ -639,53 +647,25 @@ CHIP_ERROR BLEManagerImpl::ConfigureAdvertisingData(void)
-    advertData.manufacturer_len    = 0;
-    advertData.p_manufacturer_data = NULL;
-    advertData.service_data_len    = 0;
-    advertData.p_service_data      = NULL;
-    advertData.service_uuid_len    = sizeof(UUID_CHIPoBLEService);
-    advertData.p_service_uuid      = (uint8_t *) UUID_CHIPoBLEService;
-    advertData.flag                = ESP_BLE_ADV_FLAG_GEN_DISC | ESP_BLE_ADV_FLAG_BREDR_NOT_SPT;
-    err                            = esp_ble_gap_config_adv_data(&advertData);
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(DeviceLayer, "esp_ble_gap_config_adv_data(<advertising data>) failed: %s", ErrorStr(err));
-        ExitNow();
-    }
-
-    // Construct the Chip BLE Service Data to be sent in the scan response packet.  On the ESP32,
-    // the buffer given to esp_ble_gap_config_adv_data() must begin with the Chip BLE service UUID.
-    memcpy(chipServiceData.ServiceUUID, ShortUUID_CHIPoBLEService, sizeof(chipServiceData.ServiceUUID));
-    err = ConfigurationMgr().GetBLEDeviceIdentificationInfo(chipServiceData.DeviceIdInfo);
-    SuccessOrExit(err);
-
-    // Configure the contents of the scan response packet.
-    memset(&advertData, 0, sizeof(advertData));
-    advertData.set_scan_rsp        = true;
-    advertData.include_name        = false;
-    advertData.include_txpower     = true;
-    advertData.min_interval        = 0;
-    advertData.max_interval        = 0;
-    advertData.appearance          = 0;
-    advertData.manufacturer_len    = 0;
-    advertData.p_manufacturer_data = NULL;
-    advertData.service_data_len    = sizeof(chipServiceData);
-    advertData.p_service_data      = (uint8_t *) &chipServiceData;
-    advertData.service_uuid_len    = 0;
-    advertData.p_service_uuid      = NULL;
-    advertData.flag                = 0;
-    err                            = esp_ble_gap_config_adv_data(&advertData);
+    advDataVersionDiscriminator = (discriminator | (CHIP_ADV_VERSION << 12));
+
+    memset(advData, 0, sizeof(advData));
+    advData[index++] = 0x02;                             // length
+    advData[index++] = CHIP_ADV_DATA_TYPE_FLAGS;         // AD type : flags
+    advData[index++] = CHIP_ADV_DATA_FLAGS;              // AD value
+    advData[index++] = 0x06;                             // length
+    advData[index++] = CHIP_ADV_DATA_TYPE_SERVICE_DATA;  // AD type: (Service Data - 16-bit UUID)
+    advData[index++] = ShortUUID_CHIPoBLEService[0];     // AD value
+    advData[index++] = ShortUUID_CHIPoBLEService[1];     // AD value
+    advData[index++] = CHIP_BLE_OPCODE;                  // Used to differentiate from operational CHIP Ble adv
+    advData[index++] = advDataVersionDiscriminator >> 8; // Bits[15:12] == version - Bits[11:0] Discriminator
+    advData[index++] = advDataVersionDiscriminator & 0x00FF;
+
+    // Construct the Chip BLE Service Data to be sent in the scan response packet.
+    err = esp_ble_gap_config_adv_data_raw(advData, sizeof(advData));
```

Comment:
does the size always need to be sizeof == MAX_ADV_DATA_LEN or could we use index here to reduce payload sizes? does it matter?

Reply:
At some point we will need to use manufacturer specific data and add them to the advertising data if there is enough space. The rest will go into the Scan Req/Resp packet.
I will dig into this when I will work on that in order to give you a more accurate answer.